plotly/offline/offline.py

```diff
@@ -292,7 +292,33 @@ def iplot(figure_or_data, show_link=True, link_text='Export to plot.ly',
         figure_or_data, show_link, link_text, validate,
         '100%', 525, global_requirejs=True)
 
-    display(HTML(plot_html))
+    figure = tools.return_figure_from_figure_or_data(figure_or_data, validate)
+
+    # Display 2 versions of the figure:
+    # - The raw HTML is the "old-school" style - we just throw in a JS script
+    #   inside the cell. This works for a lot of environments but some JS is
+    #   stripped out in some rendering environments / platforms like
+    #   GitHub, nteract, and eventually Plotly
+    # - The custom `mime-types` is the new-school way. nteract supports
+    #   rendering plotly graphs this way (see https://github.com/nteract/nteract/pull/662)
+    #   and others will hopefully follow suite.
+    # Both of these bundles will be saved as part of the notebook. The renderer
+    # will choose which one to render. So 'text/html' is like a fallback for
+    # any environment that doesn't have the 'application/json+plotly.v1'
+    # renderer available.
+    # The bummer is that we're injecting the figure JSON in here twice, so
+    # notebooks that are pretty big are about to get a lot bigger.
+    # The upside is that we're dropping any compatability for older notebook
+    # versions
+    display_bundle = {
+        'application/json+plotly.v1': {
+            'data': json.loads(json.dumps(figure['data'], cls=plotly.utils.PlotlyJSONEncoder)),
+            'layout': json.loads(json.dumps(figure['layout'], cls=plotly.utils.PlotlyJSONEncoder))
+        },
+        'text/html': plot_html
+    }
+
+    display(display_bundle, raw=True)
```

Comment:
A jupyter frontend falls back to a lighter media type when it doesn't support the one specified. In the case of nteract, it has an ordered list called `displayOrder`. The default displayOrder is this:

```
  'application/json',
  'application/javascript',
  'text/html',
  'image/svg+xml',
  'text/markdown',
  'text/latex',
  'image/svg+xml',
  'image/gif',
  'image/png',
  'image/jpeg',
  'application/pdf',
  'text/plain',
```

when the plotly transform is part of our app, the plotly mimetype is one of the highest priority mimetypes along with geojson and vega. If that plotly transform+mimetype combo is not available, the next richest is `text/html`.

Jupyter notebook and jupyterlab do something similar, as does nbviewer and the qtconsole.